File: backtracking.py
def find_empty(board, row=0, col=0):
    # Se ultrapassarmos as linhas, não encontramos uma célula vazia.
    if row == len(board):
        return None
    # Se encontrarmos uma célula vazia, retornamos sua posição.
    if board[row][col] == 0:
        return (row, col)

    # Avançar para a próxima célula.
    col += 1
    if col == len(board[row]):
        col = 0
        row += 1
    # Chamada recursiva para a próxima célula.
    return find_empty(board, row, col)


def valid(board, num, row, col):
    size = len(board)
    # Determinar o tamanho das regiões com base no tamanho do tabuleiro.
    box_rows, box_cols = (2, 3) if size == 6 else (
        int(size**0.5), int(size**0.5))

    # Verificar se o número é válido na linha.
    row_valid = all([board[row][i] != num for i in range(size)])
    # Verificar se o número é válido na coluna.
    col_valid = all([board[i][col] != num for i in range(size)])

    # Identificar a caixa/região à qual a célula pertence.
    box_x = col // box_cols
    box_y = row // box_rows
    # Verificar se o número é válido na caixa/região.
    box_valid = all([board[i][j] != num
                     for i in range(box_y * box_rows,
                                    (box_y + 1) * box_rows)
                     for j in range(box_x * box_cols,
                                    (box_x + 1) * box_cols)])

    # Retorna True se o número for válido em todas as verificações.
    return row_valid and col_valid and box_valid


def solve_cell(board, num, row, col):
    # Verificar se o número é válido para a célula.
    if not valid(board, num, row, col):
        return None
    # Se for válido, criar uma cópia do
    # tabuleiro e colocar o número na célula.
    new_board = [list(r) for r in board]
    new_board[row][col] = num
    # Tentar resolver o tabuleiro com o novo número.
    return solve(new_board)


def solve(board):
    # Procurar uma célula vazia.
    empty = find_empty(board)
    # Se não houver célula vazia,
    # o tabuleiro está completo.
    if not empty:
        return board
    row, col = empty
    size = len(board)
    # Tentar inserir números de 1
    # ao tamanho do tabuleiro na
    # célula vazia.
    for num in range(1, size + 1):
        # Se um número for válido,
        # continuamos resolvendo a
        # partir dessa configuração.
        solved = solve_cell(board, num, row, col)
        if solved:
            return solved
    # Se nenhum número for válido,
    # retornamos None para indicar
    # uma solução inválida.
    return None


def print_row(row, col_idx=0):
    size = len(row)
    # Determinar o tamanho das regiões com base no tamanho do tabuleiro.
    box_size = 2 if size == 6 else int(size**0.5)
    if col_idx == size:
        print()
        return
    sep = " | " if col_idx % box_size == box_size - 1 else " "
    print(row[col_idx], end=sep)
    print_row(row, col_idx + 1)


def print_board(bo, row_idx=0):
    if not bo:
        print("No solution exists for the provided Sudoku.")
        return
    size = len(bo)
    # Determinar o tamanho das regiões com base no tamanho do tabuleiro.
    box_size = 2 if size == 6 else int(size**0.5)
    if row_idx == size:
        return
    if row_idx % box_size == 0 and row_idx != 0:
        print("-" * (2 * size + box_size - 1))
    print_row(bo[row_idx])
    print_board(bo, row_idx + 1)


# Teste 4x4
board_4x4 = [
    [0, 2, 0, 0],
    [4, 0, 3, 0],
    [0, 3, 0, 4],
    [0, 0, 2, 0]
]

# Teste 6x6
board_6x6 = [
    [5, 0, 0, 6, 0, 4],
    [0, 0, 6, 0, 0, 0],
    [0, 4, 0, 0, 5, 0],
    [0, 2, 0, 0, 3, 0],
    [0, 0, 0, 4, 0, 0],
    [1, 0, 5, 0, 0, 6]
]
# Teste 9x9
board_9x9 = [
    [7, 8, 0, 4, 0, 0, 1, 2, 0],
    [6, 0, 0, 0, 7, 5, 0, 0, 9],
    [0, 0, 0, 6, 0, 1, 0, 7, 8],
    [0, 0, 7, 0, 4, 0, 2, 6, 0],
    [0, 0, 1, 0, 5, 0, 9, 3, 0],
    [9, 0, 4, 0, 6, 0, 0, 0, 5],
    [0, 7, 0, 3, 0, 0, 0, 1, 2],
    [1, 2, 0, 0, 0, 7, 4, 0, 0],
    [0, 4, 9, 2, 0, 6, 0, 0, 7]
]

print("4--------------------------------4\n")
print_board(board_4x4)
result = solve(board_4x4)
print("__________________________\n")
print_board(result)
'''
print("\n6--------------------------------6\n")
print_board(board_6x6)
result = solve(board_6x6)
print("__________________________\n")
print_board(result) '''


print("\n9--------------------------------9\n")
print_board(board_9x9)
result = solve(board_9x9)
print("__________________________\n")
print_board(result)

File: backtracking.hs
{--
1. A função solveComparative é chamada para resolver o tabuleiro.
2. Ela verifica se há alguma célula vazia usando a função findEmpty.
3. Se não houver células vazias, o tabuleiro é considerado resolvido e é retornado.
4. Se houver uma célula vazia, ela tentará preencher essa célula
   com um número válido usando a função tryNumber.
5. A função tryNumber tentará preencher a célula com um número de 1 até o tamanho do tabuleiro.
6. Para cada número, ela verifica se o número é válido usando a função isValid.
7. Se o número for válido, ela atualiza o tabuleiro e tenta
   resolver o tabuleiro atualizado chamando solveComparative recursivamente.
8. Se a tentativa de preencher a célula falhar para todos os números,
   ela retorna Nothing, indicando que não foi possível resolver o tabuleiro.
--}

import Data.List (notElem)
import Data.Maybe (fromJust)

-- Funções de Debug
import Debug.Trace-- Funções de debug

debug :: String -> a -> a
debug msg val = trace msg val

debugTryNumber :: Int -> Int -> Int -> Bool
debugTryNumber num row col = debug ("Tentando o número " ++ show num ++ " na posição (" ++ show row ++ "," ++ show col ++ ")") True

debugIsValid :: Int -> Int -> Int -> Bool
debugIsValid num row col = debug ("Número " ++ show num ++ " é válido em (" ++ show row ++ "," ++ show col ++ ")") True

debugNotValid :: Int -> Int -> Int -> Bool
debugNotValid num row col = debug ("Número " ++ show num ++ " não é válido em (" ++ show row ++ "," ++ show col ++ ")") True

debugTryFillCell :: Int -> Int -> Bool
debugTryFillCell row col = debug ("Tentando preencher a célula: (" ++ show row ++ "," ++ show col ++ ")") True

-- Função para debugar comparações
debugComparative :: Char -> Int -> Int -> Bool -> Bool
debugComparative op num comparedVal condition =
    debug (debugComparativeStr op num comparedVal condition) condition

-- Função para formatar a mensagem de debug para comparações
debugComparativeStr :: Char -> Int -> Int -> Bool -> String
debugComparativeStr op num comparedVal condition =
    "Comparing " ++ show num ++ " " ++ [op] ++ " " ++ show comparedVal ++ ": " ++ show condition

-- Fim das Funções de Debug

type Cell = (Int, Char, Char, Char, Char)
type Board = [[Cell]]

-- Função auxiliar para obter o primeiro elemento de uma célula
firstElem :: Cell -> Int
firstElem (x, _, _, _, _) = x
-- Funções auxiliares para obter partes de uma célula
getSecond (_, x, _, _, _) = x
getThird (_, _, x, _, _) = x
getFourth (_, _, _, x, _) = x
getFifth (_, _, _, _, x) = x

-- Verifica se um número é válido na linha atual
isRowValid :: Board -> Int -> Int -> Bool
isRowValid board num row = notElem num (map firstElem (board !! row))

-- Verifica se um número é válido na coluna atual
isColValid :: Board -> Int -> Int -> Bool
isColValid board num col = notElem num [firstElem (board !! r !! col) | r <- [0..length board - 1]]

-- Verifica se um número é válido na subcaixa atual
isBoxValid :: Board -> Int -> Int -> Int -> Bool
isBoxValid board num row col =
    let boxRows = if length board == 4 then 2 else if length board == 6 then 3 else 3
        boxCols = if length board == 4 then 2 else if length board == 6 then 2 else 3
        startRow = (row `div` boxRows) * boxRows
        startCol = (col `div` boxCols) * boxCols
        box = [board !! r !! c | r <- [startRow..(startRow + boxRows - 1)], c <- [startCol..(startCol + boxCols - 1)]]
    in notElem num (map firstElem box)

-- Verifica se um número é válido considerando as comparações
isComparativeValid :: Board -> Int -> Int -> Int -> Bool
isComparativeValid board num row col =
    let (_, left, up, right, down) = board !! row !! col
        leftVal  = if col == 0 then 0 else firstElem (board !! row !! (col-1))
        upVal    = if row == 0 then 0 else firstElem (board !! (row-1) !! col)
        rightVal = if col == (length (board !! 0) - 1) then 0 else firstElem (board !! row !! (col+1))
        downVal  = if row == (length board - 1) then 0 else firstElem (board !! (row+1) !! col)
    
    in debugComparative '<' num leftVal (left == '<' && num < leftVal) &&
       debugComparative '>' num leftVal (left == '>' && num > leftVal) &&
       debugComparative '<' num upVal (up == '<' && num < upVal) &&
       debugComparative '>' num upVal (up == '>' && num > upVal) &&
       debugComparative '<' num rightVal (right == '<' && num < rightVal) &&
       debugComparative '>' num rightVal (right == '>' && num > rightVal) &&
       debugComparative '<' num downVal (down == '<' && num < downVal) &&
       debugComparative '>' num downVal (down == '>' && num > downVal) &&

       (left == '/' || (left == '<' && num < leftVal) || (left == '>' && num > leftVal)) &&
       (up == '/' || (up == '<' && num < upVal) || (up == '>' && num > upVal)) &&
       (right == '/' || (right == '<' && num < rightVal) || (right == '>' && num > rightVal)) &&
       (down == '/' || (down == '<' && num < downVal) || (down == '>' && num > downVal))

-- Verifica se um número é válido considerando todas as regras
isValid :: Board -> Int -> Int -> Int -> Bool
isValid board num row col =
    let rowValid = isRowValid board num row
        colValid = isColValid board num col
        boxValid = isBoxValid board num row col
        compValid = isComparativeValid board num row col
    in if not rowValid then debug "Row invalid" False
       else if not colValid then debug "Col invalid" False
       else if not boxValid then debug "Box invalid" False
       else if not compValid then debug "Comparison invalid" False
       else True

-- Substitui um elemento em uma lista 2D
replace2D :: [[a]] -> (Int, Int) -> a -> [[a]]
replace2D matrix (i, j) x = 
    take i matrix ++ [replace (matrix !! i) j x] ++ drop (i+1) matrix
  where
    replace row j x = take j row ++ [x] ++ drop (j+1) row

-- Encontra uma célula vazia no tabuleiro
findEmpty :: Board -> Int -> Int -> Maybe (Int, Int)
findEmpty board row col
    | row == length board = Nothing
    | firstElem (board !! row !! col) == 0 = Just (row, col)
    | col == (length (board !! 0) - 1) = findEmpty board (row+1) 0
    | otherwise = findEmpty board row (col+1)

-- Tenta preencher uma célula vazia com um número válido
tryNumber :: Board -> Int -> Int -> Int -> Maybe Board
tryNumber board num row col
    | num > length board = Nothing
    | otherwise = 
        if debugTryNumber num row col then
            if isValid board num row col then
                let newCell = (num, getSecond (board !! row !! col),
                                    getThird (board !! row !! col),
                                    getFourth (board !! row !! col),
                                    getFifth (board !! row !! col))
                    newBoard = replace2D board (row, col) newCell
                in solveComparative newBoard
            else tryNumber board (num+1) row col
        else tryNumber board (num+1) row col



-- Resolve o tabuleiro
solveComparative :: Board -> Maybe Board
solveComparative board
    | findEmpty board 0 0 == Nothing = Just board
    | otherwise = 
        let (row, col) = fromJust (findEmpty board 0 0)
        in if debugTryFillCell row col then tryNumber board 1 row col
           else Nothing


-- Exibe uma célula
printCell :: Cell -> IO ()
printCell (value, _, _, _, _) = putStr (show value ++ " ")

-- Exibe o tabuleiro
printBoard :: Board -> IO ()
printBoard [] = return ()
printBoard (row:rows) = do
    mapM_ printCell row
    putStrLn ""
    printBoard rows

-- Lê o tabuleiro de uma string
readBoard :: String -> Board
readBoard input = read input :: Board

-- Função principal
main :: IO ()
main = do
    content <- readFile "tabuleiro.txt"
    let board = readBoard content
    let solution = solveComparative board
    case solution of
         Just solved -> printBoard solved
         Nothing -> putStrLn "No solution found"
